File: core/src/com/mygdx/game/characters/AI/RandomTravelAI.java
package com.mygdx.game.characters.AI;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.math.Vector3;
import com.mygdx.game.characters.GameCharacter;
import com.mygdx.game.characters.State.Stand;
import com.mygdx.game.core.ARandom;
import com.mygdx.game.ship.TileType;

import java.util.ArrayList;

/**
 * Created by al on 5/10/2016.
 */
public class RandomTravelAI extends AI {

    public float stateTime;
    public float checkTime = 0;
    Vector3 target;
    private ArrayList<Astar.ANode> available;
    private ArrayList<Astar.ANode> path;


    public RandomTravelAI(GameCharacter node) {
        super(node);
        target = null;
        checkTime = 3;
        path = new ArrayList<Astar.ANode>();
    }

    @Override
    public void update() {
        stateTime += Gdx.graphics.getDeltaTime();

        if (target == null) {
            target = node.basicCenter();
        }

        if (checkTime <= 0 && node.state.getClass() == Stand.class){

            waitRandom();

            setTarget();
        }

        travelTo();


        checkTime-= Gdx.graphics.getDeltaTime();
    }

    private void setTarget() {
        Astar astar = new Astar(node);
        available = astar.available();
        Astar.ANode targetTile = available.get(ARandom.rand((int) System.currentTimeMillis(), available.size()));
//        target = targetTile.basicCenter();

        path = astar.findTarget(targetTile.current);
        target = path.get(0).basicCenter();
    }

    public void waitRandom(){
        System.out.println("Available: " + new Astar(node).available());
//        System.out.println("Tile Position: " + node.currentTile().getCenter());

        checkTime = ARandom.rand((int) System.currentTimeMillis(), 4) + 3;

    }

    private void travelTo() {
        if (path.size() >  0){
            if(path.get(0).type == TileType.Walkable){
                walkTo();
            }else if(path.get(0).type == TileType.Climbable){
                climbTo();
            }
        }
    }

    private void climbTo() {
        Vector3 rpos = new Vector3(target);
        Vector3 npos = new Vector3(node.basicCenter());
        if (rpos.y > npos.y + 1){
            node.state.climbUp();
        } else if(rpos.y < npos.y - 1) {
            node.state.climbDown();
        }else
        {
            node.state.stand();
            path.remove(path.get(0));
            if (path.size() > 0)
                target = path.get(0).basicCenter();
        }
    }

    public void walkTo(){
        Vector3 rpos = new Vector3(target);
        Vector3 npos = new Vector3(node.basicCenter());
        if (rpos.x > npos.x + 1){
            node.state.runRight();
        } else if(rpos.x < npos.x - 1) {
            node.state.runLeft();
        }else
        {
            node.state.stand();
            path.remove(path.get(0));
            if (path.size() > 0)
                target = path.get(0).basicCenter();
        }
    }
}

File: web/src/characters/ai/RandomTravelAI.ts
import { AI } from './AI';
import { Astar, type ANode } from './Astar';
import { Stand } from '../state/Stand';
import { TileType } from '../../ship/TileType';
import { Random } from '../../core/Random';
import type { GameCharacter } from '../GameCharacter';

export class RandomTravelAI extends AI {
  private checkTime = 3;
  private target: { x: number; y: number } | null = null;
  private path: ANode[] = [];

  constructor(node: GameCharacter) {
    super(node);
  }

  override update(deltaSeconds: number): void {
    if (this.target === null) {
      this.target = this.node.basicCenter();
    }

    if (this.checkTime <= 0 && this.node.fsm instanceof Stand) {
      this.waitRandom();
      this.setTarget();
    }

    this.travelTo();
    this.checkTime -= deltaSeconds;
  }

  private setTarget(): void {
    try {
      const astar = new Astar(this.node);
      const available = astar.available();
      if (available.length === 0) return;
      const pick = available[Random.rand(Date.now(), available.length)];
      this.path = astar.findTarget(pick.current);
      if (this.path.length > 0) this.target = this.path[0].current.basicCenter();
    } catch {
      // Character not on a ship tile yet; will retry next frame.
    }
  }

  private waitRandom(): void {
    this.checkTime = Random.rand(Date.now(), 4) + 3;
  }

  private travelTo(): void {
    if (this.path.length === 0) return;
    const head = this.path[0];
    if (head.type === TileType.Walkable) {
      this.walkTo();
    } else if (head.type === TileType.Climbable) {
      this.climbTo();
    }
  }

  private walkTo(): void {
    if (!this.target) return;
    const npos = this.node.basicCenter();
    if (this.target.x > npos.x + 1) {
      this.node.fsm.runRight();
    } else if (this.target.x < npos.x - 1) {
      this.node.fsm.runLeft();
    } else {
      this.node.fsm.stand();
      this.advancePath();
    }
  }

  private climbTo(): void {
    if (!this.target) return;
    const npos = this.node.basicCenter();
    if (this.target.y > npos.y + 6) {
      this.node.fsm.climbUp();
    } else if (this.target.y < npos.y + 8) {
      this.node.fsm.climbDown();
    } else {
      this.node.fsm.stand();
      this.advancePath();
    }
  }

  private advancePath(): void {
    this.path.shift();
    if (this.path.length > 0) {
      this.target = this.path[0].current.basicCenter();
    }
  }
}
